### ISSUE TITLE
Should we add an upcall for registering a sig alg OID?

### ISSUE THREAD

**Issue by mattcaswell** [openssl/openssl#15624]
A the moment we don't have the ability to handle sig alg OIDs that we don't already know about. This means that you can't have an X509 certificate chain where intermediate or root CA have public keys that we don't already know - because we would not recognise the OID for their signature. This also means you couldn't have a self-signed cert using an algorithm we don't already know.

We actually already have the ability for application code to register their own sig alg OIDs via `OBJ_add_sigid` and `OBJ_create`

So, a provider that was linked to the same instance of libcrypto as the application could do something like the following in its `OSSL_provider_init` function:

````
int mdnid = OBJ_create("1.2.3.4", "myhash", "myhash");
int signid = OBJ_create("1.2.3.5", "mysig", "mysig");
int sigalgnid = OBJ_create("1.2.3.6", "mysigalg", "mysigalg");

OBJ_add_sigid(sigalgnid, mdnid, signid);
````

AFAICT that would be sufficient (untested) for a provider to register a new sigalg and an application could use it anywhere it likes in a cert chain.

The above fails if the provider isn't linked against the same instance of libcrypto (e.g. if static linking is being used). It could easily be wrapped up in an upcall though. Should we add that?

**Comment by paulidale** [comment]
IMO: we should.  Should we do it for 3.0?  I think not.

**Comment by levitte** [comment]
I would rather go for a mechanism to make sigalgs directly fetchable. In my view, this would be an interim measure.

**Comment by paulidale** [comment]
Which means both doesn't it?

**Comment by levitte** [comment]
What I meant, btw, is an added API to be able to use fetched EVP_SIGNATUREs directly and explicitly.

So no, it doesn't mean both, it really means that OBJ_find_sigid_algs and OBJ_find_sigid_by_algs can (and I'm of the opinion that they should) be made obsolete.

So, simply put, something like this would be suitable:

``` C
EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new(pkey, NULL);
EVP_SIGNATURE *sig = EVP_SIGNATURE_fetch(libctx, "sha1WithRSAEncryption", propq);

EVP_SIGNATURE_init(ctx, params);
EVP_SIGNATURE_update(ctx, data, datalen);
EVP_SIGNATURE_final(ctx, sig, &siglen, sigsize);
```

There's no *real* reason why anyone should *have to* do the split-oid-into-two-ids-and-rejoin-them dance.
And of course, that fetch works with un-named OIDs as well, so the fetch above can as well be:

``` C
EVP_SIGNATURE *sig = EVP_SIGNATURE_fetch(libctx, "1.2.840.113549.1.1.5", propq);
```

... as long as this is part of the algorithm names given in the OSSL_ALGORITHM.

**Comment by mattcaswell** [comment]
>  In my view, this would be an interim measure.

Yes - my suggestion was intended as an interim measure that we can do relatively easily now before the fuller solution that you outline is implemented in some future release.

**Comment by levitte** [comment]
Yeah, if it's clearly marked as an interim measure, I'm on board with that.

What troubles me is that such interim measures often become not so interim, and I worry about putting that into the provider side...  it's at the same level as adding a function that gets immediately deprecated.

So, one last question to help me decide: do you believe that there will be new sigalgs appearing anytime soon, i.e. before 3.1?
And just for the base of comparison, I plan on a serious introduction of the fuller solution as soon as possible, at the latest directly after we release of 3.0.0.  I believe it should be part of 3.1.
My thinking is that as long as our internal libcrypto table (`crypto/objects/obj_xref.h`) can be assumed to be sufficient, there's really no need to introduce any updating functionality of that table to providers.

**Comment by anhu** [comment]
Just to add a bit of context, I'm trying to write a 3.0.0 provider with signature algorithms that OpenSSL doesn't know about.

**Comment by mattcaswell** [comment]
> do you believe that there will be new sigalgs appearing anytime soon, i.e. before 3.1?

My assumption is that @baentsch and others on the oqs project would be interested in this prior to 3.1 - and we've just heard that @anhu is doing the same. So, yes, I think this would be useful for people before 3.1

**Comment by levitte** [comment]
> So, yes, I think this would be useful for people before 3.1

Okie.

**Comment by mattcaswell** [comment]
Fixed in #15681
